panther_driver/src/panther_hardware.py

```diff
@@ -0,0 +1,192 @@
+        GPIO.setup(CHRG_EN, GPIO.OUT, initial=1)
+        GPIO.setup(VDIG_OFF, GPIO.OUT, initial=0)
+        GPIO.setup(DRIVER_EN, GPIO.OUT, initial=0)
+        GPIO.setup(E_STOP_RESET, GPIO.IN)  # USED AS I/O
+
+    def _motor_start_sequence(self) -> None:
+        """
+        First start sequence for motors which is meant to power up modules in correct order
+        """
+        GPIO.output(VMOT_ON, 1)
+        time.sleep(0.5)
+
+        GPIO.output(DRIVER_EN, 1)
+        time.sleep(0.2)
+
+        GPIO.output(AUX_PW_EN, 1)
+
+    def _publish_e_stop_state(self, event=None) -> None:
+        msg = Bool()
+        msg.data = self._read_e_stop_pin()
+        self._e_stop_state_pub.publish(msg)
+
+    def _publish_charger_state(self, event=None) -> None:
+        msg = Bool()
+        msg.data = GPIO.input(CHRG_SENSE)
+        self._charger_state_pub.publish(msg)
+
+    def _aux_power_enable_cb(self, req: SetBoolRequest) -> SetBoolResponse:
+        return self._handle_set_bool_srv(req.data, AUX_PW_EN, "Aux power enable")
+
+    def _charger_enable_cb(self, req: SetBoolRequest) -> SetBoolResponse:
+        return self._handle_set_bool_srv(req.data, CHRG_EN, "Charger enable")
+
+    def _digital_power_enable_cb(self, req: SetBoolRequest) -> SetBoolResponse:
+        return self._handle_set_bool_srv(not req.data, VDIG_OFF, "Digital power enable")
+
+    def _motors_enable_cb(self, req: SetBoolRequest) -> SetBoolResponse:
+        return self._handle_set_bool_srv(req.data, DRIVER_EN, "Motors driver enable")
+
+    def _fan_enable_cb(self, req: SetBoolRequest) -> SetBoolResponse:
+        return self._handle_set_bool_srv(req.data, FAN_SW, "Fan enable")
+
+    def _e_stop_trigger_cb(self, req: TriggerRequest) -> TriggerResponse:
+        self._watchdog.turn_off()
+        return TriggerResponse(True, f"E-STOP triggered, watchdog turned off")
+
+    def _e_stop_reset_cb(self, req: TriggerRequest) -> TriggerResponse:
+        if self._validate_e_stop_pin(False):
+            return TriggerResponse(
+                False, "E-STOP was not triggered, reset is not needed"
+            )
+
+        self._reset_e_stop()
+
+        if self._validate_e_stop_pin(True):
+            return TriggerResponse(
+                False,
+                "E-STOP reset not successful, state unchanged, check for pressed E-STOP buttons or other sources"
+            )
+
+        return TriggerResponse(False, "E-STOP reset successful")
+
+    def _reset_e_stop(self) -> None:
+        GPIO.setup(E_STOP_RESET, GPIO.OUT)
+        self._watchdog.turn_on()
+
+        GPIO.output(E_STOP_RESET, True)
+        time.sleep(0.1)
+
+        GPIO.setup(E_STOP_RESET, GPIO.IN)
+
+    def _handle_set_bool_srv(self, value: bool, pin: int, name: str) -> SetBoolResponse:
+        rospy.logdebug(f"Requested {name} = {value}")
+        GPIO.output(pin, value)
+        success = self._validate_gpio_pin(pin, value)
+
+        msg = f"{name} write {value} failed"
```

Comment:
I'd keep close success and msg (remove enter) since they are creating together SetBoolResponse.

Reply:
:+1: